gpt3/Megatron-LM.patch

```diff
@@ -0,0 +1,266 @@
+ from .grad_scaler import ConstantGradScaler, DynamicGradScaler
+ from .optimizer import Float16OptimizerWithFloat16Params, FP32Optimizer
+ 
++import torch
++from msamp.optim import LBAdamW
++from msamp.megatron import FP8DistributedOptimizer as DistributedOptimizer
+ 
+ def get_param_groups(modules,
+                      no_weight_decay_cond,
+@@ -73,11 +75,21 @@ def get_megatron_optimizer(model,
+                                     lr_mult)
+ 
+     if args.optimizer == 'adam':
+-        optimizer = Adam(param_groups,
+-                         lr=args.lr,
+-                         weight_decay=args.weight_decay,
+-                         betas=(args.adam_beta1, args.adam_beta2),
+-                         eps=args.adam_eps)
++        if args.msamp:
++            exp_avg_dtype, exp_avg_sq_dtype = torch.uint8, torch.float16
++            optimizer = LBAdamW(param_groups,
++                            lr=args.lr,
++                            weight_decay=args.weight_decay,
++                            betas=(args.adam_beta1, args.adam_beta2),
++                            eps=args.adam_eps,
++                            exp_avg_dtype=exp_avg_dtype, exp_avg_sq_dtype=exp_avg_sq_dtype,
++                            tensor_scale=True)
++        else:
++            optimizer = Adam(param_groups,
++                            lr=args.lr,
++                            weight_decay=args.weight_decay,
++                            betas=(args.adam_beta1, args.adam_beta2),
++                            eps=args.adam_eps)
+     elif args.optimizer == 'sgd':
+         optimizer = SGD(param_groups,
+                         lr=args.lr,
+diff --git a/megatron/optimizer/optimizer.py b/megatron/optimizer/optimizer.py
+index da9cd70..7fb0dfb 100644
+--- a/megatron/optimizer/optimizer.py
++++ b/megatron/optimizer/optimizer.py
+@@ -13,13 +13,15 @@ from torch._utils import _flatten_dense_tensors, _unflatten_dense_tensors
+ from megatron import get_timers
+ from megatron import print_rank_0
+ from megatron.core import mpu, tensor_parallel
+-from megatron.model import DistributedDataParallel as LocalDDP
++# from megatron.model import DistributedDataParallel as LocalDDP
+ from megatron.model import Float16Module
+ from megatron.model.module import param_is_not_shared
+ from megatron.utils import unwrap_model
+ 
+-from .clip_grads import clip_grad_norm_fp32, count_zeros_fp32
++from .clip_grads import count_zeros_fp32
+ 
++from msamp.megatron import clip_grad_norm_fp8
++from msamp.megatron import FP8DistributedDataParallel as LocalDDP
+ 
+ def _zero_grad_group_helper(group, set_to_none):
+     """Zero out the gradient for a group of parameters.
+@@ -116,7 +118,7 @@ class MegatronOptimizer(ABC):
+     def clip_grad_norm(self, clip_grad):
+         params = self.get_parameters()
+         grads_for_norm = self.get_main_grads_for_grad_norm()
+-        return clip_grad_norm_fp32(
++        return clip_grad_norm_fp8(
```

Comment:
The function name `clip_grad_norm_fp8` is misleading, since it does not support FP8 only.
We can rename it to `clip_grad_norm`, `msamp_clip_grad_norm` or its original name.

Reply:
Fixed.